File: 042_trapping_rain_water/Solution.py
class Solution:
    def trap(self, height_arr):
        if not height_arr or len(height_arr) < 3: return 0

        arr_len = len(height_arr)
        left_max, right_max = height_arr[0], height_arr[arr_len - 1]
        l, r = 1, arr_len - 2
        result = 0

        while l <= r:
            if left_max <= right_max:
                result += max(0, left_max - height_arr[l])
                left_max = max(left_max, height_arr[l])
                l += 1
            else:
                result += max(0, right_max - height_arr[r])
                right_max = max(right_max, height_arr[r])
                r -= 1

        return result


if __name__ == '__main__':
    height_arr = [0, 1, 0, 2, 1, 0, 1, 3, 2, 1, 2, 1]
    solution = Solution()
    print(solution.trap(height_arr))

File: 042_trapping_rain_water/Solution.java
class Solution {
    public int trap(int[] heightArr) {
        if (heightArr == null || heightArr.length < 3) return 0;

        int arrLen = heightArr.length;
        int leftMax = heightArr[0];
        int rightMax = heightArr[arrLen - 1];
        int result = 0;
        int L = 1, R = arrLen - 2;

        while (L <= R) {
            if (leftMax <= rightMax) {
                result += Math.max(0, leftMax - heightArr[L]);
                leftMax = Math.max(leftMax, heightArr[L++]);
            } else {
                result += Math.max(0, rightMax - heightArr[R]);
                rightMax = Math.max(rightMax, heightArr[R--]);
            }
        }

        return result;
    }

    public static void main(String[] args) {
        int[] heightArr = new int[]{0, 1, 0, 2, 1, 0, 1, 3, 2, 1, 2, 1};
        Solution solution = new Solution();
        System.out.println(solution.trap(heightArr));
    }
}
